src/approaches/gnn_policy_approach.py

```diff
@@ -0,0 +1,497 @@
+            index += 1
+        for unary_predicate in unary_predicates:
+            self._node_feature_to_index[unary_predicate] = index
+            index += 1
+        for unary_predicate in unary_predicates:
+            self._node_feature_to_index[G(unary_predicate)] = index
+            index += 1
+        for obj_attr in obj_attrs:
+            self._node_feature_to_index[obj_attr] = index
+            index += 1
+
+        # Initialize input edge features.
+        self._edge_feature_to_index = {}
+        index = 0
+        for binary_predicate in binary_predicates:
+            self._edge_feature_to_index[binary_predicate] = index
+            index += 1
+        for binary_predicate in binary_predicates:
+            self._edge_feature_to_index[R(binary_predicate)] = index
+            index += 1
+        for binary_predicate in binary_predicates:
+            self._edge_feature_to_index[G(binary_predicate)] = index
+            index += 1
+        for binary_predicate in binary_predicates:
+            self._edge_feature_to_index[G(R(binary_predicate))] = index
+            index += 1
+
+    def _graphify_data(
+            self, inputs: List[Tuple[State, Set[GroundAtom], Set[GroundAtom]]],
+            targets: List[_Option]) -> Tuple[List[Dict], List[Dict]]:
+        graph_inputs = []
+        graph_targets = []
+        assert len(inputs) == len(targets)
+
+        for (state, atoms, goal), option in zip(inputs, targets):
+            # Create input graph.
+            graph_input, object_to_node = self._graphify_single_input(
+                state=state, atoms=atoms, goal=goal)
+            graph_inputs.append(graph_input)
+            # Create target graph.
+            ## First, copy over all unchanged fields.
+            graph_target = {
+                "n_node": graph_input["n_node"],
+                "n_edge": graph_input["n_edge"],
+                "edges": graph_input["edges"],
+                "senders": graph_input["senders"],
+                "receivers": graph_input["receivers"],
+            }
+            ## Next, set up the target node features.
+            object_mask = np.zeros(
+                (len(object_to_node), self._max_option_objects),
+                dtype=np.int64)
+            for i, obj in enumerate(option.objects):
+                object_mask[object_to_node[obj], i] = 1
+            graph_target["nodes"] = object_mask
+            ## Finally, set up the target globals.
+            option_index = self._sorted_options.index(option.parent)
+            onehot_target = np.zeros(len(self._sorted_options))
+            onehot_target[option_index] = 1
+            assert len(option.params.shape) == 1
+            params_target = np.zeros(self._max_option_params)
+            params_target[:option.params.shape[0]] = option.params
+            graph_target["globals"] = np.r_[onehot_target, params_target]
+            graph_targets.append(graph_target)
+
+        if CFG.gnn_policy_do_normalization:
```

Comment:
this will happen twice, first for the exemplar creation, and then for the full dataset. the second will overwrite the first, which is good, because we want to normalize with the full dataset. but maybe should clarify/fix this anyway.